pkg/storer/epoch_migration.go

```diff
@@ -0,0 +1,482 @@
+		},
+		DecodeKey: func(key []byte) (e shed.Item, err error) {
+			e.BinID = binary.BigEndian.Uint64(key[1:9])
+			return e, nil
+		},
+		EncodeValue: func(fields shed.Item) (value []byte, err error) {
+			value = make([]byte, 64) // 32 bytes address, 32 bytes batch id
+			copy(value, fields.Address)
+			copy(value[32:], fields.BatchID)
+			return value, nil
+		},
+		DecodeValue: func(keyItem shed.Item, value []byte) (e shed.Item, err error) {
+			e.Address = value[:32]
+			e.BatchID = value[32:64]
+			return e, nil
+		},
+	})
+	if err != nil {
+		return shed.Index{}, shed.Index{}, err
+	}
+
+	// Index storing actual chunk address, data and bin id.
+	headerSize := 16 + postage.StampSize
+	retrievalDataIndex, err = dbshed.NewIndex("Address->StoreTimestamp|BinID|BatchID|BatchIndex|Sig|Location", shed.IndexFuncs{
+		EncodeKey: func(fields shed.Item) (key []byte, err error) {
+			return fields.Address, nil
+		},
+		DecodeKey: func(key []byte) (e shed.Item, err error) {
+			e.Address = key
+			return e, nil
+		},
+		EncodeValue: func(fields shed.Item) (value []byte, err error) {
+			b := make([]byte, headerSize)
+			binary.BigEndian.PutUint64(b[:8], fields.BinID)
+			binary.BigEndian.PutUint64(b[8:16], uint64(fields.StoreTimestamp))
+			stamp, err := postage.NewStamp(fields.BatchID, fields.Index, fields.Timestamp, fields.Sig).MarshalBinary()
+			if err != nil {
+				return nil, err
+			}
+			copy(b[16:], stamp)
+			value = append(b, fields.Location...)
+			return value, nil
+		},
+		DecodeValue: func(keyItem shed.Item, value []byte) (e shed.Item, err error) {
+			e.StoreTimestamp = int64(binary.BigEndian.Uint64(value[8:16]))
+			e.BinID = binary.BigEndian.Uint64(value[:8])
+			stamp := new(postage.Stamp)
+			if err = stamp.UnmarshalBinary(value[16:headerSize]); err != nil {
+				return e, err
+			}
+			e.BatchID = stamp.BatchID()
+			e.Index = stamp.Index()
+			e.Timestamp = stamp.Timestamp()
+			e.Sig = stamp.Sig()
+			e.Location = value[headerSize:]
+			return e, nil
+		},
+	})
+	if err != nil {
+		return shed.Index{}, shed.Index{}, err
+	}
+
+	return pullIndex, retrievalDataIndex, nil
+}
+
+// epochMigrator is a helper struct for migrating epoch data. It is used to house
+// the main logic of the migration so that it can be tested. Also it houses the
+// dependencies of the migration logic.
+type epochMigrator struct {
+	stateStore         storage.StateStorer
+	store              storage.Store
+	recovery           *sharky.Recovery
+	reserve            *reserve.Reserve
+	pullIndex          shed.Index
+	retrievalDataIndex shed.Index
+	logger             log.Logger
+}
+
+func (e *epochMigrator) migrateReserve(ctx context.Context) error {
+	type putOp struct {
+		pIdx  shed.Item
+		chunk swarm.Chunk
+		loc   sharky.Location
+	}
+
+	e.logger.Debug("migrating reserve contents")
+
+	opChan := make(chan putOp, 4)
+	eg, egCtx := errgroup.WithContext(ctx)
+
+	for i := 0; i < 4; i++ {
+		eg.Go(func() error {
+			for {
+				select {
+				case <-ctx.Done():
```

Comment:
There is no need to handle the `ctx` here, `egCtx` is its descendant, so context cancellation is propagated to it. The same is below.